(WIP) CRM-16243 - Extension API - Automatically enable dependencies

totten opened the pull request:
This applies to the `install` and `enable` actions.

Handling `download` is a different task.  Ideally, we'd use `composer` for
downloading. But I'm not sure if that needs to block support for `install`/`enable`.

This consumes dependency data from `info.xml`, as in:

```xml
<extension key="org.civicrm.cvform" type="module">
  <requires>
    <ext>org.civicrm.api4</ext>
  </requires>
</extension>
```

Consequently, one would need to declare extension dependencies twice (once for
the download dependencies in `composer.json`; once for the runtime dependencies in `info.xml`).

It has received very negligible acceptance testing.  Please feel free to
nitpick away!

---

 * [CRM-16243: Dependency management for extensions](https://issues.civicrm.org/jira/browse/CRM-16243)

totten commented:
Note: I flagged as `WIP` not because there's any particular change outstanding... but it might benefit some more experimenting/thinking about different workflows.

colemanw commented:
Not that this has to be implemented now, just want to make sure we're not painting ourselves into a corner with the proposed xml syntax. If we want to start supporting things like versioned dependencies, would it perhaps look like:

```
<extension key="org.civicrm.cvform" type="module">
  <requires>
    <ext>org.civicrm.api4</ext>
    <min>1.0</min>
    <max>1.2</max>
  </requires>
</extension>
```

totten commented:
Generally: pea-shoe, the topsort library is PHP 5.4 only. Probably easiest to fork it and patch for PHP 5.3.

@colemanw , if we want to put version constraints in there, then we should probably make it part of the `<ext>` tag so that it works with multiple requirements, e.g.

```xml
  <requires>
    <ext min="1.0" max="1.2">org.civicrm.api4</ext>
   <!-- or -->
    <ext version="~1.2.0">org.civicrm.api4</ext>
   <!-- or -->
    <ext>org.civicrm.api4@~1.2.0</ext>
  </requires>
```

However, I'm not sure we should. It turns out that there are several differences in how you handle dependency-resolution for purposes of *downloading* vs *activating*.
 * When downloading, you need to make a decision about which versions to fetch, and hopefully you fetch the right versions. Both `composer` and `drush make` have fairly established semantics about which versions they'll download. Also, you need to resolve dependencies in the space of *local/existing* packages as well as *remote/available* packages.
 * When activating, you've already got all the code; you can't change versions, and it's a waste of time to check *remote/available* packages. However, there's an additional constraint -- you should activate requirements in the right *order*. (Ex: If the `cvform` module uses `api4` during installation, then it's best to activate `api4` first.) (Note: Neither `composer` nor `drush make` handles activation. They leave that others.)

So... versions matter when it comes to downloads. But I don't really want to maintain our current downloader (`CRM_Extension_Browser`/`CRM_Extension_Downloader`) -- there are lots of details/edge-cases in downloading that `composer` handles better.

My main hesitation about this PR is that (longer term) dependencies would be articulated in both `info.xml`  and `composer.json` -- better to have them in one place. I'm not sure how to do that without either forcing changes on downstream devs or doing a good chunk of additional dev. However, I don't think we'd be unique in having this situation, and it's not that hard to do...

totten commented:
Updated to use PHP 5.3-compatible variant of `topsort`. Rebased/squashed.

Compatibility fixes sent upstream ( https://github.com/marcj/topsort.php/pull/2 ).

totten commented:
If you have a downloader (like `composer install`), then you can enable all the downloaded extensions by a path, e.g.

```
cv api extension.install path=$PWD/vendor/*
```

(More in commit notes).

totten commented on CRM/Extension/Manager.php:
```diff
@@ -292,6 +292,13 @@ public function disable($keys) {
+      throw new CRM_Extension_Exception_DependencyException("Cannot disable extension due dependencies. Consider disabling all these: " . implode(',', $disableRequirements));
+    }
+
     foreach ($keys as $key) {
```
FIXME: This `foreach` should use the topological ordering returned by `findDisableRequirements()`. As written, it uses the user-supplied ordering.

totten commented:
jenkins, test this please

totten commented:
jenkins, test this please

jaapjansma commented:
@totten I have setup a test account for this patch and it fails. I have added requirements section to the info.xml of https://github.com/CiviCooP/org.civicoop.civiruleslogger and I could install that extension without any failure, although I have uninstalled and removed the dependent extensions.

So it looks like this patch is not working at all. I also miss the UI which explains to the user which extensions are required and which requirements are met. But at least a failure and an error messages during install would be nice.

totten commented:
@jaapjansma , thanks for testing. It would help to have a slightly more verbose description of the test procedure -- e.g. this is what I inferred from your comments:

```
civibuild create pr8022 --type drupal-clean --patch https://github.com/civicrm/civicrm-core/pull/8022
cd build/pr8022/sites/all/modules/civicrm/tools/extensions
git clone https://github.com/CiviCooP/org.civicoop.civiruleslogger
## Assert: The source code for org.civicoop.civirules does NOT exist.
drush cvapi extension.refresh
drush cvapi extension.install key=org.civicoop.civiruleslogger
```

One would expect this to fail because a dependency (`org.civicoop.civirules`) is uninstallable. From the description, it sounds like the above process incorrectly enabled `civiruleslogger` on your system.

However, when I run the scenario, I do get a correct error (albeit terse):

```
drush cvapi extension.install key=org.civicoop.civiruleslogger
Array
(
    [is_error] => 1
    [error_message] => Unknown extension: org.civicoop.civirules
)
```

Since we got different results, perhaps I misinterpreted the steps?

jaapjansma commented:
@totten you are right my test description could be better. This is what I did:

1. On the command line `civibuild create pr8022 --type drupal-clean --patch https://github.com/civicrm/civicrm-core/pull/8022`
2. In the browser: Login into the civicrm installations and disable and then uninstall _org.civicoop.civirules_
3. On the command line: `cd build/pr8022/sites/all/modules/civicrm/tools/extensions `
4 .On the command line `rm -rf org.civicoop.civirules`
4. On the command line `git clone https://github.com/CiviCooP/org.civicoop.civiruleslogger`
5. Back to the browser login into civicrm and go the manage extensions
6. Check whether _org.civicoop.civirules_ is present or not. It should not be present.
7. Enable and install the _org.civicoop.civiruleslogger_ extension

After step 7 the _org.civicoop.civiruleslogger_ extension is installed without any warning or failure to the user.

totten commented:
@jaapjansma , thank you, that's quite helpful. :) You're right - my PR has totally omitted UI considerations.

Notes to future-self for fixing this:
 * Check whether the admin UI calls the Extension API. It should probably use API (or else call `$manager->findInstallRequirements($keys)`).
 * Need to update UI to *display* the dependencies to admins.

omarabuhussein commented on CRM/Extension/Info.php:
```diff
@@ -141,6 +168,12 @@ public function parse($info) {
           );
         }
       }
+      elseif ($attr === 'requires') {
```
i think it would more neat if parse() method organized in such way instead of this list of if/else :

```php
    foreach ($info as $attr => $val) {
      if (count($val->children()) == 0) {
        $this->$attr = (string) $val;
        continue;
      }

      switch ($attr) {
        case 'urls':
          # code...
          break;
        case 'classloader':
          # code...
          break;
        case 'requires':
          # code...
          break;
        default:
          # code...
          break;
      }
    }
```

guanhuan commented:
Will it be better to also have versions in dependency management. e.g. instead of having
`<requires>
    <ext>org.civicrm.api4</ext>
  </requires>`
We can have
`<requires>
    <ext><key>org.civicrm.api4</key><ver>2.0beta</ver></ext>
  </requires>`
Dependency is mostly version sensitive and if we have the version in the specification, we could automatically download dependencies for the published extensions as well (which encourages people make their work more visible).

omarabuhussein commented:
> Consequently, one would need to declare extension dependencies twice (once for
the download dependencies in composer.json; once for the runtime dependencies in info.xml).

Our team discussed this internally , and the first thing we said ... WHY ?? :D  ...

I mean it is not elegant to define dependencies twice ... 

Basically what came to my mind are three ways to handle downloads : 

1- simply you don't handle them , you just check if they are existing in extension directory and install them if needed , otherwise you just throw an error that installing this extension require ( a list of dependencies ) which I think is already handled by this PR .

2- you handle downloads via civicrm downloader  ( which you pointed that is something you don't want to maintain and that's why you are thinking of using composer ) , the problem of this approach other than that maintaining   the downloader  is there is a need for a central place/server that hold extensions data and all extensions should be registered there , but I think this is already exist ( https://civicrm.org/extensions ) . 

3- a suggestion by @davialexandre that is to use composer  but to hide it's usage , so the extension is defined only once in info.xml file , and internally civicrm will use composer to handle downloading dependencies  without the need for the developer to set the dependencies in composer too manually.

Personally I find the 2nd option is the best among them , I Don't think that a lot of work need to be done to make the civirm downloader works for most cases and I don't think it will need a lot of maintaining ( But indeed I may be mistaken ) .

eileenmcnaughton commented:
@totten I'm closing this as it is stale but please re-open if it comes back to life

totten commented:
To respond/recap a few things:

 * The PR lingered in review/WIP because it needed a patch so that the core UI used it (as pointed out by Jaap).
 * Regarding downloads, I believe this should be viewed as an orthogonal problem. 
   * I do like the idea of `composer` downloads. The reason is discussed more in http://civicrm.stackexchange.com/questions/11357/how-to-develop-extension-that-relies-on-libs-brought-in-by-composer . The great thing about `composer` for this is that they've defined behaviors for all kinds of edge-cases (version conflicts; downloading via different media; package overrides; private listings; caching; etal). But there's a flipside: there are more edge-cases where `composer` will spit out unexpected messages+prompts, so that makes writing a wrapper more involved. Even so, I feel that moving toward `composer` is the best value for anyone who can work with CLI.
   * While we should obviously hope for a consensus way to handle downloading of the codebases (there are real contenders floating around -- e.g. `composer` vs `drush make` vs `git submodules` vs general sysadmin tools), this PR does not require a consensus. Regardless of how code is obtained, the PR addresses the activation.

eileenmcnaughton commented:
@totten so should I understand this should be re-opened & merged as an improvement but not a full solution?
